src/middlewared/middlewared/plugins/reporting.py

```diff
@@ -0,0 +1,951 @@
+        ('cache_size-arc', 'value', None),
+        ('cache_size-L2', 'value', None),
+    )
+
+    def get_title(self):
+        return 'ARC Size'
+
+
+class ARCRatioPlugin(RRDBase):
+
+    plugin = 'zfs_arc'
+    vertical_label = 'Hits (%)'
+    rrd_types = (
+        ('cache_ratio-arc', 'value', '%name%,100,*'),
+        ('cache_ratio-L2', 'value', '%name%,100,*'),
+    )
+
+    def get_title(self):
+        return 'ARC Hit Ratio'
+
+
+class ARCResultPlugin(RRDBase):
+
+    identifier_plugin = False
+    plugin = 'zfs_arc'
+    vertical_label = 'Requests'
+    rrd_data_extra = """
+        CDEF:total=%name_0%,%name_1%,+
+        XPORT:total:total
+    """
+
+    def get_rrd_types(self, identifier):
+        return (
+            (f'cache_result-{identifier}-hit', 'value', '%name%,100,*'),
+            (f'cache_result-{identifier}-miss', 'value', '%name%,100,*'),
+        )
+
+    def get_title(self):
+        return 'ARC Requests ({identifier})'
+
+    def get_identifiers(self):
+        return ('demand_data', 'demand_metadata', 'prefetch_data', 'prefetch_metadata')
+
+
+class NFSStatPlugin(RRDBase):
+
+    plugin = 'nfsstat-client'
+    title = 'NFS Stats'
+    vertical_label = 'Bytes'
+    rrd_types = (
+        ('nfsstat-read', 'value', None),
+        ('nfsstat-write', 'value', None),
+    )
+
+
+class ReportingService(Service):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.__rrds = {}
+        for name, klass in RRD_PLUGINS.items():
+            self.__rrds[name] = klass(self.middleware)
+
+    @filterable
+    def graphs(self, filters, options):
+        return filter_list([i.__getstate__() for i in self.__rrds.values()], filters, options)
+
+    def __rquery_to_start_end(self, query):
+        unit = query.get('unit')
+        if unit:
+            verrors = ValidationErrors()
+            for i in ('start', 'end'):
+                if i in query:
+                    verrors.add(
+                        f'reporting_query.{i}',
+                        f'{i!r} should only be used if "unit" attribute is not provided.',
+                    )
+            verrors.check()
+        else:
+            if 'start' not in query:
+                unit = 'HOURLY'
+            else:
+                starttime = query['start']
+                endtime = query.get('end') or 'now'
+
+        if unit:
+            unit = unit[0].lower()
+            page = query['page']
+            starttime = f'end-{page + 1}{unit}'
+            if not page:
+                endtime = 'now'
+            else:
+                endtime = f'now-{page}{unit}'
+        return starttime, endtime
+
+    @accepts(
+        List('graphs', items=[
+            Dict(
+                'graph',
+                Str('name', required=True),
+                Str('identifier', default=None, null=True),
+            ),
+        ], empty=False),
+        Dict(
+            'reporting_query',
+            Str('unit', enum=[
```

Comment:
`unit` is a unit for `page`, like, `page=6 unit=hours`, why is it an adverb then?

Reply:
What do you propose instead?

Reply:
`["HOUR", "DAY", ...]`